src/lib/core/hooks/variableCoverage.ts

```diff
@@ -0,0 +1,112 @@
+import { useMemo } from 'react';
+
+import { CompleteCasesTable } from '../api/data-api';
+import {
+  VariableCoverageTableRow,
+  VariableSpec,
+} from '../components/VariableCoverageTable';
+import { Filter } from '../types/filter';
+
+import { useEntityCounts } from './entityCounts';
+
+/**
+ * Returns an array of data to be rendered in a VariableCoverageTable
+ * @param completeCases
+ * @param filteredEntityCounts
+ * @param variables
+ * @param outputEntityId
+ * @returns A VariableCoverageTableRow[]
+ */
+export function useVariableCoverageTableRows(
+  variableSpecs: VariableSpec[],
+  filters: Filter[],
+  completeCases?: CompleteCasesTable,
+  outputEntityId?: string
+): VariableCoverageTableRow[] {
+  const completeCasesMap = useCompleteCasesMap(completeCases);
+  const filteredEntityCountsResult = useEntityCounts(filters);
+
+  return useMemo(
+    () =>
+      variableSpecs.map((spec) => {
+        if (spec.variable == null) {
+          return {
+            role: spec.role,
+            required: spec.required,
+          };
+        }
+
+        const { entityId, variableId } = spec.variable;
+
+        const variableCompleteCases =
+          completeCasesMap[`${entityId}.${variableId}`];
+
+        if (variableCompleteCases == null) {
+          return {
+            role: spec.role,
+            required: spec.required,
+            display: spec.display,
+          };
+        }
+
+        const baseRowWithCounts = {
+          role: spec.role,
+          required: spec.required,
+          display: spec.display,
+          completeCount: variableCompleteCases,
+        };
+
+        const variableFilteredEntityCount =
+          filteredEntityCountsResult.value == null || outputEntityId == null
+            ? undefined
+            : filteredEntityCountsResult.value[outputEntityId];
+
+        if (variableFilteredEntityCount == null) {
+          return baseRowWithCounts;
+        }
+
+        const incompleteCount =
+          variableFilteredEntityCount - baseRowWithCounts.completeCount;
+
+        return {
+          ...baseRowWithCounts,
+          incompleteCount,
+          completePercent:
+            (baseRowWithCounts.completeCount / variableFilteredEntityCount) *
+            100,
+          incompletePercent:
+            (incompleteCount / variableFilteredEntityCount) * 100,
+        };
+      }),
+    [
+      completeCasesMap,
+      filteredEntityCountsResult.value,
+      outputEntityId,
+      variableSpecs,
+    ]
+  );
+}
+
+/**
+ * Transforms a CompleteCasesTable array into a map
+ * @param completeCases
+ * @returns A map with key-value pairs of the form {entityId}.{variableId} -> {completeCaseCount}
+ */
+export function useCompleteCasesMap(
+  completeCases: CompleteCasesTable = []
+): Record<string, number> {
+  return useMemo(
+    () =>
+      completeCases.reduce((memo, { completeCases, variableDetails }) => {
```

Comment:
Is this really a reduction?  If the same `variableDetails.variableId` is seen twice, the former value will be silently overwritten.

It seems that `lodash.keyBy` would do something similar, but it won't clean up the array-or-not-array completeCases mess, as shown in my toy example.

```
a = [ { name: 'john', age: 23 }, { name: 'mary', age: [ 22 ] } ]
z = _.keyBy(a, (obj) => obj.name)
{
  john: { name: 'john', age: 23 },
  mary: { name: 'mary', age: [ 22 ] }
}
```

Reply:
Though `_.keyBy` also silently overwrites!  I just expect more from `reduce` ;-)

Reply:
You've kicked up two important issues:

1.  To resolve the "overwriting issue," we could update the backend so that each "row" of the `completeCases` response includes not just the id of the variable, but also the variable's role (xAxisVariable, yAxisVariable, etc). With such a change, the complete cases map could be keyed by variable role instead of variable id.

2. While you're right that replacing my `reduce` with a `keyBy` won't automatically resolve the "overwriting issue" or the "array-or-not-array completeCases mess," there's still a clear opportunity to use `keyBy` to improve readability:

Given the complete cases rows:

* `filter` away the rows that are missing `completeCases` or `variableDetails`
* `keyBy` the variable id associated to each row
* `map` each row to its complete cases count (via lodash's `mapValues`)

That means making three passes through the data (instead of one pass with the `reduce` implementation), but that's OK. If these three passes are a source of performance issues, we have much bigger issues on our hands!

Reply:
Well, it's an interesting case of readability vs conciseness and performance.

I hadn't thought of the incomplete cases part (which makes it a proper reduction, I had missed that) - so it ends up being six of one and half a dozen of the other (though I learned something in the process).